localstack/http/proxy.py

```diff
@@ -0,0 +1,151 @@
+        """
+        Creates a new Handler.
+
+        :param forward_path_parameter: if set to true, the forwarder will use the `path` argument from the keyword
+            arguments passed to the endpoint to forward, s.t. you can add rules like ("/foo/<path:path>") that will
+            implicitly strip `/foo` from the request.
+        """
+        self.proxy = proxy
+        self.forward_path_parameter = forward_path_parameter
+
+    def __call__(self, request: Request, **kwargs) -> Response:
+        if self.forward_path_parameter:
+            forward_path = kwargs.pop("path", "")
+        else:
+            forward_path = request.path
+
+        return self.proxy.forward(request, forward_path=forward_path)
+
+
+def path_forwarder(forward_base_url: str) -> ForwardingHandler:
+    """
+    Creates a new Handler that can be attached to a Router that uses a handler dispatcher for forwarding requests to
+    a given destination. This can be a very convenient tool to create simple forwarding rules.
+
+    For instance, if you want to forward all requests from /foobar/<path:path> to http://localhost:8080/v1/<path>,
+    you would do the following::
+
+        router = Router(dispatcher=handler_dispatcher())
+        router.add("/foobar/<path:path>", path_forwarder("http://localhost:8080/v1")
+
+    This is similar to the common nginx configuration where proxy_pass is a URI::
+
+        location /foobar {
+            proxy_pass http://localhost:8080/v1/;
+        }
+
+    :param forward_base_url: the base url to forward to
+    :return: a new ForwardingHandler with the appropriate configuration
+    """
+    return ForwardingHandler(Proxy(forward_base_url=forward_base_url), forward_path_parameter=True)
+
+
+def _copy_request(
+    request: Request,
+    base_url: str = None,
+    path: str = None,
+    headers: Union[Headers, Mapping[str, str]] = None,
+) -> Request:
+    """
+    Creates a new request from the given one that can be used to perform a proxy call.
+
+    :param request: the original request
+    :param base_url: the url to forward the request to (e.g., http://localhost:8080)
+    :param path: the path to forward the request to (e.g., /foobar), if set to None, the original path will be used
+    :param headers: optional headers to overwrite
+    :return: a new request with slightly modified underlying environment but the same data stream
+    """
+    builder = EnvironBuilder.from_environ(request.environ)
+
+    if base_url:
+        builder.base_url = base_url
+        builder.headers["Host"] = builder.host
+
+    if path is not None:
+        builder.path = path
+
+    if headers:
+        builder.headers.update(headers)
+
+    # FIXME: unfortunately, EnvironBuilder expects the input stream to be seekable, but we don't have that when using
+    #  the asgi/wsgi bridge. we need a better way of dealing with IO!
+    builder.input_stream = BytesIO(restore_payload(request))
```

Comment:
q: Maybe we could tackle this problem in a future iteration by wrapping the incoming request's stream with our own stream (a bit like we do in the content decoding handler) that allows us to access the original content easily afterwards?

Reply:
yes good idea! we can work out the details together at some point.